File: packages/flutter/lib/src/face_tracker.dart
import 'dart:math';
import 'types.dart';
import 'face_geometry.dart';

// ── Constants ──────────────────────────────────────────────────────────────────
const double trackMatchThreshold     = 0.20;
const int    trackStaleMs            = 3000;
const int    trackConfirmMs          = 300;
const int    boxDisplayMs            = 150;
const double trackSelectMinFrontalness  = 40;
const double pendingConfirmThreshold    = 50;
const int    minPendingConfirmedFrames  = 3;

// ── TrackedFace ────────────────────────────────────────────────────────────────
class TrackedFace {
  TrackedFace({
    required this.id,
    required this.createdAt,
    required double cx,
    required double cy,
    required double nW,
    required double nH,
  })  : cx = cx,
        cy = cy,
        dispCx = cx,
        dispCy = cy,
        dispW = nW,
        dispH = nH,
        bestCx = cx,
        bestCy = cy,
        pendingCx = cx,
        pendingCy = cy,
        pendingBw = nW,
        pendingBh = nH,
        lastSeenAt = DateTime.now().millisecondsSinceEpoch;

  final String id;
  final int    createdAt;

  double cx, cy;
  double dispCx, dispCy, dispW, dispH;
  String? bestJpeg;
  double  bestCx, bestCy;
  double  bestFrontalness = 0;
  String? selectedJpeg;
  double  selectedFrontalness = 0;
  double  currentFrontalness  = 0;
  double  smoothedFrontalness = 0;
  int     lastSeenAt;
  int     lastSentAt          = 0;
  double  lastSentFrontalness = 0;
  String? pendingJpeg;
  double  pendingFrontalness  = 0;
  double  pendingCx, pendingCy;
  double  pendingBw, pendingBh;
  int     pendingStaleFrames      = 0;
  int     pendingConfirmedFrames  = 0;
}

int _trackIdCounter = 0;

TrackedFace createTrack(
  double cx, double cy, double nW, double nH, int nowMs,
) {
  return TrackedFace(
    id:        (++_trackIdCounter).toString(),
    createdAt: nowMs,
    cx: cx, cy: cy, nW: nW, nH: nH,
  );
}

// ── FrameBuffer interface ─────────────────────────────────────────────────────

abstract class FrameBuffer {
  int    get width;
  int    get height;
  String toJpegBase64({double quality = 0.85});
  String? cropFaceJpeg(double cx, double cy, double bw, double bh, {double quality = 0.85});
}

// ── Commit pipeline ────────────────────────────────────────────────────────────

SelectedFace _toSelectedFace(TrackedFace t) => SelectedFace(
  trackId:    t.id,
  dataUrl:    t.selectedJpeg!,
  frontalness: t.selectedFrontalness,
  lastSentAt: DateTime.fromMillisecondsSinceEpoch(t.lastSentAt),
);

void _resetPending(TrackedFace track) {
  track.pendingFrontalness     = 0;
  track.pendingJpeg            = null;
  track.pendingStaleFrames     = 0;
  track.pendingConfirmedFrames = 0;
}

void commitPending(
  TrackedFace track,
  void Function(SelectedFace) onSelect,
) {
  if (track.pendingFrontalness > track.bestFrontalness) {
    track.bestFrontalness = track.pendingFrontalness;
  }
  track.bestJpeg = track.pendingJpeg;
  track.bestCx   = track.pendingCx;
  track.bestCy   = track.pendingCy;

  if (track.pendingFrontalness > track.selectedFrontalness ||
      track.selectedJpeg == null) {
    track.selectedFrontalness = track.pendingFrontalness;
    track.selectedJpeg        = track.pendingJpeg;
    if (track.selectedJpeg != null &&
        track.pendingFrontalness >= trackSelectMinFrontalness) {
      onSelect(_toSelectedFace(track));
    }
  }
  _resetPending(track);
}

// ── Track management ───────────────────────────────────────────────────────────

void updateTracks(
  List<Detection> detections,
  FrameBuffer frame,
  List<TrackedFace> tracks,
  void Function(TrackedFace) onCommit,
) {
  final nowMs    = DateTime.now().millisecondsSinceEpoch;
  final available = List<TrackedFace>.from(tracks);

  for (final det in detections) {
    final box = det.boundingBox;
    final cx  = box.originX + box.width  / 2;
    final cy  = box.originY + box.height / 2;
    final nW  = box.width;
    final nH  = box.height;

    int    bestIdx  = -1;
    double bestDist = double.infinity;
    for (var i = 0; i < available.length; i++) {
      final dist = sqrt(
        pow(cx - available[i].cx, 2) + pow(cy - available[i].cy, 2),
      );
      if (dist < trackMatchThreshold && dist < bestDist) {
        bestDist = dist; bestIdx = i;
      }
    }

    TrackedFace track;
    if (bestIdx == -1) {
      track = createTrack(cx, cy, nW, nH, nowMs);
      tracks.add(track);
    } else {
      track = available[bestIdx];
      available.removeAt(bestIdx);
    }

    track
      ..cx        = cx
      ..cy        = cy
      ..lastSeenAt = nowMs
      ..dispCx    = cx
      ..dispCy    = cy
      ..dispW     = nW
      ..dispH     = nH;

    final frontalness = getFrontalness(det.keypoints, nW, nH) * det.score;
    track.currentFrontalness  = frontalness;
    track.smoothedFrontalness = track.smoothedFrontalness == 0
        ? frontalness
        : track.smoothedFrontalness * 0.7 + frontalness * 0.3;

    final confirmed = nowMs - track.createdAt >= trackConfirmMs;
    if (confirmed) {
      if (frontalness >= pendingConfirmThreshold) track.pendingConfirmedFrames++;

      if (track.smoothedFrontalness > track.pendingFrontalness ||
          track.pendingJpeg == null) {
        track.pendingFrontalness = track.smoothedFrontalness;
        track.pendingJpeg        =
            frame.cropFaceJpeg(cx, cy, nW, nH) ?? frame.toJpegBase64();
        track.pendingCx          = cx;
        track.pendingCy          = cy;
        track.pendingBw          = nW;
        track.pendingBh          = nH;
        track.pendingStaleFrames = 0;
      } else {
        track.pendingStaleFrames++;
        if (track.pendingStaleFrames >= 2 && track.pendingJpeg != null) {
          onCommit(track);
        }
      }
    }
  }

  for (final track in available) {
    track.pendingStaleFrames++;
    if (track.pendingStaleFrames >= 2 &&
        track.pendingFrontalness > track.bestFrontalness) {
      onCommit(track);
    }
  }
}

List<TrackedFace> flushStale(
  List<TrackedFace> tracks,
  void Function(TrackedFace) onCommit,
  void Function(String) onRemove,
) {
  final nowMs   = DateTime.now().millisecondsSinceEpoch;
  final removed = tracks.where((t) => nowMs - t.lastSeenAt >= trackStaleMs).toList();
  for (final t in removed) {
    if (t.pendingJpeg != null) onCommit(t);
    onRemove(t.id);
  }
  return tracks.where((t) => nowMs - t.lastSeenAt < trackStaleMs).toList();
}

File: packages/core/src/face-tracker.ts
import { getFrontalness } from "./face-geometry";
import type { Detection, FrameBuffer, SelectedFace } from "./types";

// ── Constants ──────────────────────────────────────────────────────────────────
export const TRACK_MATCH_THRESHOLD        = 0.20;  // max normalised centroid distance
export const TRACK_STALE_MS               = 3000;  // drop track if unseen this long
export const TRACK_CONFIRM_MS             = 300;   // face must be tracked before SELECTED
export const BOX_DISPLAY_MS               = 150;   // hide overlay box after last seen
export const TRACK_SELECT_MIN_FRONTALNESS = 40;    // min frontalness to fire onSelect
export const PENDING_CONFIRM_THRESHOLD    = 50;    // min frontalness to count a confirmed frame
export const MIN_PENDING_CONFIRMED_FRAMES = 3;     // reject single-frame MediaPipe spikes
export const TRACK_COOLDOWN_MS             = 10_000; // min ms between sends for the same track

// ── Types ──────────────────────────────────────────────────────────────────────
export interface TrackedFace {
  id: string;
  createdAt: number;
  cx: number; cy: number;
  dispCx: number; dispCy: number; dispW: number; dispH: number;
  bestJpeg: string | null;
  bestCx: number; bestCy: number;
  bestFrontalness: number;
  selectedJpeg: string | null;
  selectedFrontalness: number;
  currentFrontalness: number;
  smoothedFrontalness: number;
  lastSeenAt: number;
  lastSentAt: number;
  lastSentFrontalness: number;
  pendingJpeg: string | null;
  pendingFrontalness: number;
  pendingCx: number;
  pendingCy: number;
  pendingBw: number;
  pendingBh: number;
  pendingStaleFrames: number;
  pendingConfirmedFrames: number;
}

// ── Factory ────────────────────────────────────────────────────────────────────
let trackIdCounter = 0;

export function createTrack(
  cx: number, cy: number,
  nW: number, nH: number,
  nowMs: number,
): TrackedFace {
  return {
    id: String(++trackIdCounter),
    createdAt: nowMs, cx, cy,
    dispCx: cx, dispCy: cy, dispW: nW, dispH: nH,
    bestJpeg: null, bestCx: cx, bestCy: cy, bestFrontalness: 0,
    selectedJpeg: null, selectedFrontalness: 0,
    currentFrontalness: 0, smoothedFrontalness: 0,
    lastSeenAt: nowMs, lastSentAt: 0, lastSentFrontalness: 0,
    pendingJpeg: null, pendingFrontalness: 0,
    pendingCx: cx, pendingCy: cy, pendingBw: nW, pendingBh: nH,
    pendingStaleFrames: 0, pendingConfirmedFrames: 0,
  };
}

// ── Helpers ────────────────────────────────────────────────────────────────────
function toSelectedFace(t: TrackedFace): SelectedFace {
  return {
    trackId: t.id,
    dataUrl: t.selectedJpeg!,
    frontalness: t.selectedFrontalness,
    lastSentAt: t.lastSentAt,
    createdAt: t.createdAt,
  };
}

function resetPending(track: TrackedFace): void {
  track.pendingFrontalness = 0;
  track.pendingJpeg = null;
  track.pendingStaleFrames = 0;
  track.pendingConfirmedFrames = 0;
  track.smoothedFrontalness = 0;
}

// ── Commit pipeline ────────────────────────────────────────────────────────────

/**
 * Promote the pending peak to the confirmed best.
 * Calls onSelect only if quality passes the minimum threshold.
 */
export function commitPending(
  track: TrackedFace,
  onSelect: (face: SelectedFace) => void,
): void {
  if (track.pendingFrontalness > track.bestFrontalness) {
    track.bestFrontalness = track.pendingFrontalness;
  }
  track.bestJpeg = track.pendingJpeg;
  track.bestCx   = track.pendingCx;
  track.bestCy   = track.pendingCy;

  if (
    track.pendingFrontalness > track.selectedFrontalness ||
    track.selectedJpeg === null
  ) {
    track.selectedFrontalness = track.pendingFrontalness;
    track.selectedJpeg        = track.pendingJpeg;
    if (track.selectedJpeg && track.pendingFrontalness >= TRACK_SELECT_MIN_FRONTALNESS) {
      track.lastSentAt = Date.now();
      onSelect(toSelectedFace(track));
    }
  }
  resetPending(track);
}

// ── Track management ───────────────────────────────────────────────────────────

/**
 * Match detected faces to existing tracks, create new tracks for unmatched
 * faces, update pending capture state, and trigger commits when a peak is
 * confirmed. Mutates `tracks` (push only).
 */
export function updateTracks(
  detections: Detection[],
  frame: FrameBuffer,
  tracks: TrackedFace[],
  onCommit: (track: TrackedFace) => void,
): void {
  const nowMs    = Date.now();
  const available = [...tracks];

  for (const det of detections) {
    const box = det.boundingBox;
    // Convert absolute pixel coords → normalised (detections are already normalised)
    const cx = box.originX + box.width  / 2;
    const cy = box.originY + box.height / 2;
    const nW = box.width;
    const nH = box.height;

    let bestIdx = -1, bestDist = Infinity;
    for (let i = 0; i < available.length; i++) {
      const dist = Math.hypot(cx - available[i]!.cx, cy - available[i]!.cy);
      if (dist < TRACK_MATCH_THRESHOLD && dist < bestDist) {
        bestDist = dist; bestIdx = i;
      }
    }

    let track: TrackedFace;
    if (bestIdx === -1) {
      track = createTrack(cx, cy, nW, nH, nowMs);
      tracks.push(track);
    } else {
      track = available[bestIdx]!;
      available.splice(bestIdx, 1);
    }

    track.cx = cx; track.cy = cy; track.lastSeenAt = nowMs;
    track.dispCx = cx; track.dispCy = cy; track.dispW = nW; track.dispH = nH;

    const frontalness =
      getFrontalness(det.keypoints, nW, nH) * det.score;
    track.currentFrontalness = frontalness;
    track.smoothedFrontalness =
      track.smoothedFrontalness === 0
        ? frontalness
        : track.smoothedFrontalness * 0.7 + frontalness * 0.3;

    const confirmed = nowMs - track.createdAt >= TRACK_CONFIRM_MS;
    if (confirmed) {
      if (frontalness >= PENDING_CONFIRM_THRESHOLD) track.pendingConfirmedFrames++;

      if (
        track.smoothedFrontalness > track.pendingFrontalness ||
        track.pendingJpeg === null
      ) {
        track.pendingFrontalness = track.smoothedFrontalness;
        track.pendingJpeg        = frame.cropFaceJpeg(cx, cy, nW, nH) ?? frame.toJpegBase64();
        track.pendingCx          = cx;
        track.pendingCy          = cy;
        track.pendingBw          = nW;
        track.pendingBh          = nH;
        track.pendingStaleFrames = 0;
      } else {
        track.pendingStaleFrames++;
        const cooldownOk = nowMs - track.lastSentAt >= TRACK_COOLDOWN_MS;
        const qualityOk  = track.pendingConfirmedFrames >= MIN_PENDING_CONFIRMED_FRAMES;
        if (track.pendingStaleFrames >= 10 && track.pendingJpeg !== null && cooldownOk && qualityOk) {
          onCommit(track);
        }
      }
    }
  }

  // Advance stale countdown for unmatched tracks
  for (const track of available) {
    track.pendingStaleFrames++;
    const cooldownOk = nowMs - track.lastSentAt >= TRACK_COOLDOWN_MS;
    const qualityOk  = track.pendingConfirmedFrames >= MIN_PENDING_CONFIRMED_FRAMES;
    if (
      track.pendingStaleFrames >= 10 &&
      track.pendingFrontalness > track.bestFrontalness &&
      cooldownOk && qualityOk
    ) {
      onCommit(track);
    }
  }
}

/** Remove tracks unseen for TRACK_STALE_MS. Commits any dangling pending first. */
export function flushStale(
  tracks: TrackedFace[],
  onCommit: (track: TrackedFace) => void,
  onRemove: (id: string) => void,
): TrackedFace[] {
  const nowMs   = Date.now();
  const removed = tracks.filter((t) => nowMs - t.lastSeenAt >= TRACK_STALE_MS);
  for (const t of removed) {
    if (t.pendingJpeg) onCommit(t);
    onRemove(t.id);
  }
  return tracks.filter((t) => nowMs - t.lastSeenAt < TRACK_STALE_MS);
}
